src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ThrowsTaglet.java

```diff
@@ -25,274 +25,635 @@
+                utils.configuration.getMessages().warning(ch.getDocTreePath(tag), "doclet.inheritDocNoDoc", n);
+                if (!add) {
+                    exceptionSection.endEntry();
+                }
+                return;
+            }
+            // if {@inheritDoc} is the only tag in the @throws description and
+            // this call can add new entries to the exception section,
+            // so can the recursive call
+            boolean addNewEntryRecursively = addNewEntry && !add;
+            var tags = expansion.get();
+            if (!addNewEntryRecursively && tags.size() > 1) {
+                // current tag has more to description than just {@inheritDoc}
+                // and thus cannot expand to multiple tags;
+                // it's likely a documentation error
+                utils.configuration.getMessages().error(utils.getCommentHelper(holder).getDocTreePath(tag), "doclet.inheritDocWithinInappropriateTag");
+                return;
+            }
+            for (Map.Entry<ThrowsTree, ExecutableElement> e : tags.entrySet()) {
+                outputAnExceptionTagDeeply(exceptionSection, originalExceptionType, e.getKey(), e.getValue(), addNewEntryRecursively, alreadyDocumentedExceptions, typeSubstitutions, writer, utils);
+            }
+            // this might be an empty list, which is fine
+            if (!loneInheritDoc) {
+                Content afterInheritDoc = writer.commentTagsToOutput(holder, description.subList(i + 1, description.size()));
+                exceptionSection.continueEntry(afterInheritDoc);
+            }
+            if (add) {
+                exceptionSection.endEntry();
+            }
+        }
+    }
+
+    private static int indexOfInheritDoc(ThrowsTree tag, ExecutableElement holder)
+            throws Failure.InvalidMarkup
+    {
+        var tags = tag.getDescription();
+        int i = -1;
+        for (var iterator = tags.listIterator(); iterator.hasNext(); ) {
+            DocTree t = iterator.next();
+            if (t.getKind() == DocTree.Kind.INHERIT_DOC) {
+                if (i != -1) {
+                    throw new Failure.InvalidMarkup(t, holder);
+                }
+                i = iterator.previousIndex();
+            }
+        }
+        return i;
+    }
+
+    private static Element getExceptionType(ThrowsTree tag,
+                                            ExecutableElement holder,
+                                            Utils utils)
+            throws Failure.ExceptionTypeNotFound,
+                   Failure.NotExceptionType
+    {
+        Element e = utils.getCommentHelper(holder).getException(tag);
+        if (e == null) {
+            throw new Failure.ExceptionTypeNotFound(tag, holder);
+        }
+        var t = e.asType();
+        var subtypeTestInapplicable = t.getKind() == TypeKind.EXECUTABLE
+                || t.getKind() == TypeKind.PACKAGE
+                || t.getKind() == TypeKind.MODULE;
```

Comment:
Just asking ... would this be more concise as an enhanced switch expression?

Reply:
A `switch` expression might look uglier than you think it would. Care to provide an example? Performance implications aside, perhaps an `EnumSet` could be more readable.

Reply:
A `switch` expression might look uglier than you think it would. Care to suggest the exact substitution? Performance implications aside, perhaps an `EnumSet` could be more readable.

Reply:
A `switch` expression will likely look uglier than you might think. Care to suggest such an expression? Performance implications aside, perhaps an `EnumSet` could be more readable.